rust/ip-packet/src/lib.rs

```diff
@@ -0,0 +1,377 @@
+pub mod make;
+
+pub use pnet_packet::*;
+
+use pnet_packet::{
+    icmpv6::MutableIcmpv6Packet,
+    ip::{IpNextHeaderProtocol, IpNextHeaderProtocols},
+    ipv4::{Ipv4Packet, MutableIpv4Packet},
+    ipv6::{Ipv6Packet, MutableIpv6Packet},
+    tcp::{MutableTcpPacket, TcpPacket},
+    udp::{MutableUdpPacket, UdpPacket},
+};
+use std::net::IpAddr;
+
+macro_rules! for_both {
+    ($this:ident, |$name:ident| $body:expr) => {
+        match $this {
+            Self::Ipv4($name) => $body,
+            Self::Ipv6($name) => $body,
+        }
+    };
+}
+
+// no std::mem::swap? no problem
+macro_rules! swap_src_dst {
+    ($p:expr) => {
+        let src = $p.get_source();
+        let dst = $p.get_destination();
+        $p.set_source(dst);
+        $p.set_destination(src);
+    };
+}
+
+#[derive(Debug, PartialEq)]
+pub enum IpPacket<'a> {
+    Ipv4(Ipv4Packet<'a>),
+    Ipv6(Ipv6Packet<'a>),
+}
+
+#[derive(Debug, PartialEq)]
+pub enum MutableIpPacket<'a> {
+    Ipv4(MutableIpv4Packet<'a>),
+    Ipv6(MutableIpv6Packet<'a>),
+}
+
+impl<'a> MutableIpPacket<'a> {
+    pub fn new(buf: &'a mut [u8]) -> Option<Self> {
+        match buf[0] >> 4 {
+            4 => Some(MutableIpPacket::Ipv4(MutableIpv4Packet::new(buf)?)),
+            6 => Some(MutableIpPacket::Ipv6(MutableIpv6Packet::new(buf)?)),
+            _ => None,
+        }
+    }
+
+    pub fn owned(data: Vec<u8>) -> Option<MutableIpPacket<'static>> {
+        let packet = match data[0] >> 4 {
+            4 => MutableIpv4Packet::owned(data)?.into(),
+            6 => MutableIpv6Packet::owned(data)?.into(),
+            _ => return None,
+        };
+
+        Some(packet)
+    }
+
+    pub fn to_owned(&self) -> MutableIpPacket<'static> {
+        match self {
+            MutableIpPacket::Ipv4(i) => MutableIpv4Packet::owned(i.packet().to_vec())
+                .expect("owned packet is still valid")
+                .into(),
+            MutableIpPacket::Ipv6(i) => MutableIpv6Packet::owned(i.packet().to_vec())
+                .expect("owned packet is still valid")
+                .into(),
```

Comment:
```suggestion
            MutableIpPacket::Ipv4(i) => MutableIpv4Packet::owned(i.packet().to_vec())
                .expect("owned packet should still be valid")
                .into(),
            MutableIpPacket::Ipv6(i) => MutableIpv6Packet::owned(i.packet().to_vec())
                .expect("owned packet should still be valid")
                .into(),
```

Otherwise it will give the error "owned packet is still valid" which is ambiguous until you look at the code.

e.g. https://godbolt.org/z/KMc7zsK4n

```
Program returned: 101
Program stderr

thread 'main' panicked at /app/example.rs:3:17:
to be Some
```

Per https://doc.rust-lang.org/std/error/index.html#common-message-styles

> focus on the word “should” as in “env variable should be set by blah”